docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+        new_fee.assert_valid();
+        self.internal_set_fee(new_fee);
+    }
+}
+```
+
+## Use `log!`
+
+Use logging for debugging and notifying user.
+
+When you need a formatted message, you can use the following macro:
+
+```rust
+log!("Transferred {} tokens from {} to {}", amount, sender_id, receiver_id);
+```
+
+It's equivalent to the following message:
+
+```rust
+env::log_str(format!("Transferred {} tokens from {} to {}", amount, sender_id, receiver_id).as_ref());
+```
+
+## Return `Promise`
+
+If your method makes a cross-contract call, you probably want to return the newly created `Promise`.
+This allows the caller (such as a near-cli or near-api-js call) to wait for the result of the promise instead of returning immediately.
+Additionally, if the promise fails for some reason, returning it will let the caller know about the failure, as well as enabling NEAR Explorer and other tools to mark the whole transaction chain as failing.
+This can prevent false-positives when the first or first few transactions in a chain succeed but a subsequent transaction fails.
+
+E.g.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn withdraw_100(&mut self, receiver_id: AccountId) -> Promise {
+        Promise::new(receiver_id).transfer(100)
+    }
+}
+```
+
+## Use high-level cross-contract API
+
+There is a helper macro that allows you to make cross-contract calls with the syntax `#[ext_contract(...)]`. It takes a Rust Trait and
+converts it to a module with static methods. Each of these static methods takes positional arguments defined by the Trait,
+then the `receiver_id`, the attached deposit and the amount of gas and returns a new `Promise`.
+
+For example, let's define a calculator contract Trait:
+
+```rust
+#[ext_contract(ext_calculator)]
+trait Calculator {
+    fn mult(&self, a: U64, b: U64) -> U128;
+
+    fn sum(&self, a: U128, b: U128) -> U128;
+}
+```
+
+It's equivalent to the following code:
+
+```rust
+mod ext_calculator {
+    pub fn mult(a: U64, b: U64, receiver_id: &AccountId, deposit: Balance, gas: Gas) -> Promise {
+        Promise::new(receiver_id.clone())
+            .function_call(
+                b"mult",
+                json!({ "a": a, "b": b }).to_string().as_bytes(),
+                deposit,
+                gas,
+            )
+    }
+
+    pub fn sum(a: U128, b: U128, receiver_id: &AccountId, deposit: Balance, gas: Gas) -> Promise {
+        // ...
+    }
+}
+```
+
+Let's assume the calculator is deployed on `calc.near`, we can use the following:
+
+```rust
+const CALCULATOR_ACCOUNT_ID: &str = "calc.near";
+const NO_DEPOSIT: Balance = 0;
+const BASE_GAS: Gas = 5_000_000_000_000;
+
+#[near_bindgen]
+impl Contract {
+    pub fn sum_a_b(&mut self, a: U128, b: U128) -> Promise {
+        let calculator_account_id: AccountId = CALCULATOR_ACCOUNT_ID.to_string();
+        ext_calculator::sum(a, b, &calculator_account_id, NO_DEPOSIT, BASE_GAS)
+    }
+}
+```
```

Comment:
Should we not just have this information in the cross contract calls section?

Reply:
Thx, moved it to the section I think fits best, although there may be some shaping later in terms of how we explain traits. For now I think it reads best where I put it in the callbacks section. https://github.com/near/sdk-docs/pull/37/commits/127cd0431a50dcb0be475b1b573ed1fe48c53928